graphql_compiler/cost_estimation/interval.py

```diff
@@ -0,0 +1,91 @@
+
+    def is_empty(self) -> bool:
+        """Return whether the interval is empty"""
+        if self.lower_bound is None or self.upper_bound is None:
+            return False
+        return self.lower_bound > self.upper_bound
+
+    def __eq__(self, other) -> bool:
+        """Compare two intervals. Empty intervals are considered equal to each other."""
+        if self.is_empty() and other.is_empty():
+            return True
+        return self.lower_bound == other.lower_bound and self.upper_bound == other.upper_bound
+
+
+def measure_interval(interval: Interval[int]) -> Optional[int]:
+    if interval.lower_bound is None or interval.upper_bound is None:
+        return None
+    if interval.is_empty():
+        return 0
+    return interval.upper_bound - interval.lower_bound + 1
+
+
+def _get_stronger_lower_bound(lower_bound_a, lower_bound_b):
+    """Return the larger bound of the two given lower bounds.
+
+    Args:
+        lower_bound_a: int or None, describing one of the lower bounds. If the bound does not exist,
+                       the argument value should be None.
+        lower_bound_b: int or None, describing one of the lower bounds. If the bound does not exist,
+                       the argument value should be None.
+
+    Returns:
+        - int, the larger of the two lower bounds, if one or more lower bounds have an integer
+          value.
+        - None if both lower bounds have a value of None.
+    """
+    stronger_lower_bound = None
+    if lower_bound_a is not None and lower_bound_b is not None:
+        stronger_lower_bound = max(lower_bound_a, lower_bound_b)
+    elif lower_bound_a is not None:
+        stronger_lower_bound = lower_bound_a
+    elif lower_bound_b is not None:
+        stronger_lower_bound = lower_bound_b
+
+    return stronger_lower_bound
+
+
+def _get_stronger_upper_bound(upper_bound_a, upper_bound_b):
+    """Return the smaller bound of the two given upper bounds.
+
+    Args:
+        upper_bound_a: int or None, describing one of the upper bounds. If the bound does not exist,
+                       the argument value should be None.
+        upper_bound_b: int or None, describing one of the upper bounds. If the bound does not exist,
+                       the argument value should be None.
+
+    Returns:
+        - int, the smaller of the two upper bounds, if one or more upper bounds have an integer
+          value.
+        - None if both upper bounds have a value of None.
+    """
+    stronger_upper_bound = None
+    if upper_bound_a is not None and upper_bound_b is not None:
+        stronger_upper_bound = min(upper_bound_a, upper_bound_b)
+    elif upper_bound_a is not None:
+        stronger_upper_bound = upper_bound_a
+    elif upper_bound_b is not None:
+        stronger_upper_bound = upper_bound_b
+
+    return stronger_upper_bound
+
+
+def intersect_intervals(interval_a, interval_b):
+    """Return the intersection of two IntegerIntervals."""
```

Comment:
`IntegerIntervals` needs to be updated here, and this function needs to be generic.